### PR TITLE
Implement Asp.Net Core 3.0/3.1 generator (#6009)

### PR DESCRIPTION
**Pull request by phatcher**
Addresses #6009, mainly new templates but some minor changes to the aspnetcore generator:

- Fixed some issues with the 2.x templates which meant the resulting code did not compile.
- Improve reporting, i.e. code shows correct .Net Core version
- Introduced new templates for 3.x using idiomatic Asp.Net Core 3.
- Change default startup to show the swagger definition rather than a non-existent page.
- Improved the Docker file so that we can integrate into Visual Studio docker generation; not running under Docker at the moment

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.  @jimschubert, @frankyjuang, @shibayan

### CODE REVIEW

**Comment by phatcher** [comment]
The build failures don't appear to be anything to do with the code changes, do I need to base this off a different commit?

**Comment by wing328** [comment]
I've updated the documentation via https://github.com/OpenAPITools/openapi-generator/pull/6025/commits/fff4da94f22b9f74b6aa89c86ac4a5438719c4fd. Let's see how it goes.

**Comment by wing328** [comment]
cc @A-Joshi as well.

**Comment by phatcher** [comment]
Seems like the Swift4 tests are the culprit now

File: modules/openapi-generator/src/main/resources/aspnetcore/3.0/Startup.mustache [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        public Startup(IHostEnvironment env)
+        {
+            var builder = new ConfigurationBuilder()
+                .SetBasePath(env.ContentRootPath)
+                .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
+                .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
+                .AddEnvironmentVariables();
+
+            Configuration = builder.Build();
+        }
```
Comment (shibayan) [inline modules/openapi-generator/src/main/resources/aspnetcore/3.0/Startup.mustache:44]:
```suggestion
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }
```

File: modules/openapi-generator/src/main/resources/aspnetcore/3.0/Startup.mustache [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                    c.RouteTemplate = "swagger/{documentName}/openapi.json";
+                })
+                .UseSwaggerUI(c =>
+                {
+                    //TODO: Either use the SwaggerGen generated Swagger contract (generated from C# classes)
+                    c.SwaggerEndpoint("/swagger/{{#version}}{{{version}}}{{/version}}{{^version}}v1{{/version}}/openapi.json", "{{#appName}}{{{appName}}}{{/appName}}{{^appName}}{{packageName}}{{/appName}}");
+
+                    //TODO: Or alternatively use the original Swagger contract that's included in the static files
+                    // c.SwaggerEndpoint("/openapi-original.json", "{{#appName}}{{{appName}}}{{/appName}}{{^appName}}{{packageName}}{{/appName}} Original");
+                }){{/useSwashbuckle}};
```
Comment (shibayan) [inline modules/openapi-generator/src/main/resources/aspnetcore/3.0/Startup.mustache:162]:
```suggestion
            app.UseDefaultFiles();
            app.UseStaticFiles();
{{#useSwashbuckle}}
            app.UseSwagger(c =>
                {
                    c.RouteTemplate = "swagger/{documentName}/openapi.json";
                })
                .UseSwaggerUI(c =>
                {
                    //TODO: Either use the SwaggerGen generated Swagger contract (generated from C# classes)
                    c.SwaggerEndpoint("/swagger/{{#version}}{{{version}}}{{/version}}{{^version}}v1{{/version}}/openapi.json", "{{#appName}}{{{appName}}}{{/appName}}{{^appName}}{{packageName}}{{/appName}}");

                    //TODO: Or alternatively use the original Swagger contract that's included in the static files
                    // c.SwaggerEndpoint("/openapi-original.json", "{{#appName}}{{{appName}}}{{/appName}}{{^appName}}{{packageName}}{{/appName}} Original");
                }){{/useSwashbuckle}};
```

File: modules/openapi-generator/src/main/resources/aspnetcore/3.0/wwwroot/web.config [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (shibayan) [inline modules/openapi-generator/src/main/resources/aspnetcore/3.0/wwwroot/web.config:1]: Please delete this file.

Comment (phatcher) [reply modules/openapi-generator/src/main/resources/aspnetcore/3.0/wwwroot/web.config:1]: Should I remove the entire www directory as that's not part of the template anymore?

Comment (shibayan) [reply modules/openapi-generator/src/main/resources/aspnetcore/3.0/wwwroot/web.config:1]: Yes. It should be fine to remove it.

File: modules/openapi-generator/src/main/resources/aspnetcore/2.1/Startup.mustache [hunk 4]
```diff
@@ -44,17 +44,13 @@ namespace {{packageName}}
...
             services.AddTransient<IAuthorizationHandler, ApiKeyRequirementHandler>();
             services.AddAuthorization(authConfig =>
             {
-{{/-first}}            
                 authConfig.AddPolicy("{{name}}",
                     policyBuilder => policyBuilder
                         .AddRequirements(new ApiKeyRequirement(new[] { "my-secret-key" },"{{name}}")));
-{{#-last}}
             });
-{{/-last}}
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/aspnetcore/2.1/Startup.mustache:57]: @phatcher thanks for the PR. I noticed these `-first`, `-last` tags are removed. Would it still work if the openapi spec contains multiple API key definitions?

Comment (phatcher) [reply modules/openapi-generator/src/main/resources/aspnetcore/2.1/Startup.mustache:57]: Not sure, I removed them as the code did not compile with a single API key, there were missing "}" from a number of files,  e.g. the PetStore example, is there some conditional code missing from the aspnetcore generator?.

Is there an example with multiple API key definitions so I can check what affect -first/-last have?

Comment (wing328) [reply modules/openapi-generator/src/main/resources/aspnetcore/2.1/Startup.mustache:57]: OK. I think we can deal with it later as single API key is a lot more common.

**Comment by wing328** [comment]
If no further feedback, I'll merge this over the weekend.

**Comment by shibayan** [comment]
Please wait for the merge until web.config under wwwroot is removed. That setting is harmful.

**Comment by wing328** [comment]
@shibayan no worry. I'll remove it after merging the PR into master

**Comment by wing328** [comment]
Filed https://github.com/OpenAPITools/openapi-generator/pull/6147 to skip web.config generation.

**Comment by wing328** [comment]
To promote the new enhancement:

- https://twitter.com/oas_generator/status/1257524719231746049
- https://www.reddit.com/r/dotnet/comments/gdqsdn/autogenerate_aspnet_core_servers_stub_for_rest/
- https://news.ycombinator.com/item?id=23076379

**Comment by wing328** [comment]
@phatcher thanks for the PR, which has been included in the v4.3.1 release: https://twitter.com/oas_generator/status/1258057660974329861

**Pull request merged**